model/jpa/src/main/java/org/keycloak/grant/jpa/impl/GrantServiceImpl.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2016 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.grant.jpa.impl;
+
+import org.keycloak.models.*;
+import org.keycloak.grant.jpa.UserGrantEntity;
+import org.keycloak.models.jpa.entities.UserConsentEntity;
+import org.keycloak.storage.StorageId;
+
+import javax.persistence.EntityManager;
+import javax.persistence.LockModeType;
+import javax.persistence.TypedQuery;
+import java.util.List;
+
+public class GrantServiceImpl implements GrantService {
+
+    private final KeycloakSession session;
+    protected EntityManager em;
+
+    public GrantServiceImpl(KeycloakSession session, EntityManager em) {
+        this.session = session;
+        this.em = em;
+        if (getRealm() == null) {
+            throw new IllegalStateException("The service cannot accept a session without a realm in its context.");
+        }
+    }
+
+    protected RealmModel getRealm() {
+        return session.getContext().getRealm();
+    }
+
+
+    @Override
+    public boolean revokeGrantByGrantId(RealmModel realm, String grantId, String userId) {
+        UserGrantEntity userGrant = getUserGrantEntity(grantId, userId, LockModeType.PESSIMISTIC_WRITE);
+        if (userGrant == null) return false;
+
+        em.remove(userGrant);
+        em.flush();
+        return true;
```

Comment:
If a grant is revoked, is an access token and refresh token that have been issued when this grant has been created still active or not?
If not, a client could still use these tokens even if its related grant was revoked.

It might be better to clarify the relationship among
existing Consent, existing UserSession, and newly introduced Grant. 

If a grant is created, modified or revoked, what happens to a consent and a user session.
If a consent is created, modified or revoked, what happens to a grant.
If a user session is created, modified or revoked, what happens to a grant.

Reply:
Normally when a grant is revoked, an access token and refresh token that has been issued when this grant has been created should no longer be active. Still looking at how to implement that.

IMO  If we delete/modify a grant, there is no impact on consent neither userSession.
if we delete/modify userSession, there is no impact on a grant
if we delete a consent, all grant with the same client and user should be deleted

WDYT?

Reply:
@guymoyo Generally, I agree with you.